cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -439,6 +469,325 @@ struct ScalarMinMax {
+            num_valid++;
+          } else {
+            valid_cols[col] = util::string_view();
+          }
+        } else {
+          const ArrayData& array = *batch[col].array();
+          if (!array.MayHaveNulls() ||
+              BitUtil::GetBit(array.buffers[0]->data(), array.offset + row)) {
+            const offset_type* offsets = array.GetValues<offset_type>(1);
+            const uint8_t* data = array.GetValues<uint8_t>(2, /*absolute_offset=*/0);
+            const int64_t length = offsets[row + 1] - offsets[row];
+            valid_cols[col] = util::string_view(
+                reinterpret_cast<const char*>(data + offsets[row]), length);
+            num_valid++;
+          } else {
+            valid_cols[col] = util::string_view();
+          }
+        }
+      }
+
+      if (num_valid < batch.values.size() && !options.skip_nulls) {
+        // We had some nulls
+        builder.UnsafeAppendNull();
+        continue;
+      }
+      util::string_view result = valid_cols.front();
+      for (size_t col = 1; col < batch.values.size(); ++col) {
+        util::string_view value = valid_cols[col];
+        if (value.empty()) {
+          DCHECK(options.skip_nulls);
+          continue;
+        }
+        result = result.empty() ? value : Op::CallBinary(result, value);
+      }
+      if (result.empty()) {
+        builder.UnsafeAppendNull();
+      } else {
+        builder.UnsafeAppend(result);
+      }
+    }
+
+    std::shared_ptr<Array> string_array;
+    RETURN_NOT_OK(builder.Finish(&string_array));
+    *out = *string_array->data();
+    out->mutable_array()->type = batch[0].type();
+    DCHECK_EQ(batch.length, out->array()->length);
+    DCHECK_GE(final_size,
+              checked_cast<const ArrayType&>(*string_array).total_values_length());
+    return Status::OK();
+  }
+
+  // Compute the length of the output for the given position, or -1 if it would be null.
+  static int64_t CalculateRowSize(const ElementWiseAggregateOptions& options,
+                                  const ExecBatch& batch, const int64_t index) {
+    const auto num_args = batch.values.size();
+    int64_t final_size = 0;
+    for (size_t i = 0; i < num_args; i++) {
+      int64_t element_size = 0;
+      bool valid = true;
+      if (batch[i].is_scalar()) {
+        const Scalar& scalar = *batch[i].scalar();
+        valid = scalar.is_valid;
+        element_size = UnboxScalar<Type>::Unbox(scalar).size();
+      } else {
+        const ArrayData& array = *batch[i].array();
+        valid = !array.MayHaveNulls() ||
+                BitUtil::GetBit(array.buffers[0]->data(), array.offset + index);
+        const offset_type* offsets = array.GetValues<offset_type>(1);
+        element_size = offsets[index + 1] - offsets[index];
+      }
+      if (!valid) {
+        if (options.skip_nulls) {
+          continue;
+        }
+        return -1;
+      }
+      final_size = std::max(final_size, element_size);
```

Comment:
I forgot to add a comment here, but this is intended. My understanding is that we only need the size in order to avoid reallocation, so we can safely overestimate it. Computing the precise value would double the amount of comparisons, which I was reluctant to do for strings. Should I do that instead?

Reply:
Overallocating sounds better than doubling comparisons, but please rename variables and methods in addition to commenting and make it clear that this is an overestimate. For example, this was being used as the length for output scalars in https://github.com/apache/arrow/commit/4c4ce0ee41f5a5b9123a4c581d6f0c1fd11aa5e8#diff-368ab7e748bd4432c92d9fdc26b51e131742b968e3eb32a6fcea4b9f02fa36aaR667

Reply:
Done.